### PR TITLE
chore: run to checkpoint migration

### PR DESCRIPTION
**Pull request by NicholasBlaskey**
## Description

Adds a one run to many checkpoint relationship table. Didn't just add a field so in the future if we want to make the relationship many to many we can just drop the unique constraint on the relationship table.

## Test Plan

should be covered by integration tests

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by ioga — APPROVED** [review]
lgtm

File: master/internal/api_trials.go [hunk 1]
```diff
@@ -1385,9 +1385,22 @@ func (a *apiServer) ReportCheckpoint(
 	default:
 	}
 
-	if err := db.AddCheckpointMetadata(ctx, c); err != nil {
+	task, err := db.TaskByID(ctx, model.TaskID(req.Checkpoint.TaskId))
+	if err != nil {
+		return nil, fmt.Errorf("looking up task to decide if trial: %w", err)
+	}
+	if task.TaskType != model.TaskTypeTrial {
```
Comment (stoksc) [inline master/internal/api_trials.go:1392]: i think we have, but i would triple check there isn't a customer doing batch inference with commands and checkpointing.

**Review by stoksc — APPROVED** [review]
lgtm. meta question though: what's exactly the migration plan to get sdks submitting run IDs to `ReportCheckpoint` and it using them instead of task ids? are we going to just have a `ReportCheckpointV2` or migrate what exists? or is this not in scope to figure out just yet.

**Comment by NicholasBlaskey** [comment]
> lgtm. meta question though: what's exactly the migration plan to get sdks submitting run IDs to `ReportCheckpoint` and it using them instead of task ids? are we going to just have a `ReportCheckpointV2` or migrate what exists? or is this not in scope to figure out just yet.

I'm not sure to be honest. I think making a ```ReportCheckpointV2``` makes sense to me.

File: master/internal/api_trials.go (hunk 1 above)
Comment (NicholasBlaskey) [reply master/internal/api_trials.go:1392]: We broke this use case in https://github.com/determined-ai/determined/blob/337174635035a5edeab956ae23cd979274f271ec/master/static/migrations/20231011104015_add-exp-trials-on-del-casc.tx.up.sql#L47

It isn't too much work to fix it here I wrote a test for it here but if we broke support for it I don't see why this change should add back in support for it.
https://github.com/determined-ai/determined/pull/8535/files#diff-3ae51e125f14c5b44fa5541d9ecb8f57332daf504ec62068d141e82e8c742fe8R590

**Pull request merged**